Wrong deserialization result when deserializing a POJO type containing double fields using `FuryInputStream` with the initial buffer size < the serialized buffer size of a single record

### Search before asking

- [x] I had searched in the [issues](https://github.com/apache/fury/issues) and found no similar issues.


### Version

- Fury: both 0.8.0 and 0.9.0
- JDK: 8u442
- OS: macOS 15.3

When deserializing a POJO type containing double fields using `FuryInputStream` with the initial buffer size < the serialized buffer size of a single record, the deserialized value is wrong. Increasing the initial buffer size resolves the issue.

### Component(s)

Java

### Minimal reproduce step


- POJO class to serialize
```java
public class SimpleType {

	// Changing this to an int value won't produce wrong result
	public double dVal;

	public SimpleType() {
		dVal = 0.5;
	}
}
```

- Test: 
```java
Fury fury = Fury.builder()
		.withLanguage(Language.JAVA)
		.build();
fury.register(SimpleType.class);

SimpleType v = new SimpleType();
// Size: 12
System.out.println(fury.serialize(v).length);
ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
fury.serialize(outputStream, v);

InputStream inputStream = new ByteArrayInputStream(outputStream.toByteArray());
// <11: wrong result
// >= 12: correct result
FuryInputStream input = new FuryInputStream(inputStream, 11);
SimpleType vout = (SimpleType) fury.deserialize(input);

// java.lang.AssertionError: 
// Expected :0.5
// Actual   :1.8227805048890994E-304
Assert.assertEquals(v.dVal, vout.dVal, 0.001);
```

### What did you expect to see?

The deserialized value = 0.5.

### What did you see instead?

The deserialized value = 1.8227805048890994E-304.

### Anything Else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

--- a/java/fury-core/src/main/java/org/apache/fury/builder/ObjectCodecBuilder.java
+++ b/java/fury-core/src/main/java/org/apache/fury/builder/ObjectCodecBuilder.java
@@ -625,14 +625,14 @@ public class ObjectCodecBuilder extends BaseObjectCodecBuilder {
     List<Expression> expressions = new ArrayList<>();
     int numPrimitiveFields = getNumPrimitiveFields(primitiveGroups);
     Literal totalSizeLiteral = Literal.ofInt(totalSize);
+    // After this check, following read can be totally unsafe without checks
+    expressions.add(new Invoke(buffer, "checkReadableBytes", totalSizeLiteral));
     Expression heapBuffer =
         new Invoke(buffer, "getHeapMemory", "heapBuffer", PRIMITIVE_BYTE_ARRAY_TYPE);
     Expression readerAddr =
         new Invoke(buffer, "getUnsafeReaderAddress", "readerAddr", PRIMITIVE_LONG_TYPE);
     expressions.add(heapBuffer);
     expressions.add(readerAddr);
-    // After this check, following read can be totally unsafe without checks
-    expressions.add(new Invoke(buffer, "checkReadableBytes", totalSizeLiteral));
     int acc = 0;
     for (List<Descriptor> group : primitiveGroups) {
       ListExpression groupExpressions = new ListExpression();
@@ -690,16 +690,17 @@ public class ObjectCodecBuilder extends BaseObjectCodecBuilder {
       Expression bean, Expression buffer, List<List<Descriptor>> primitiveGroups) {
     List<Expression> expressions = new ArrayList<>();
     int numPrimitiveFields = getNumPrimitiveFields(primitiveGroups);
-    Expression heapBuffer =
-        new Invoke(buffer, "getHeapMemory", "heapBuffer", PRIMITIVE_BYTE_ARRAY_TYPE);
-    expressions.add(heapBuffer);
     for (List<Descriptor> group : primitiveGroups) {
+      // After this check, following read can be totally unsafe without checks.
+      // checkReadableBytes first, `fillBuffer` may create a new heap buffer.
+      ReplaceStub checkReadableBytesStub = new ReplaceStub();
+      expressions.add(checkReadableBytesStub);
+      Expression heapBuffer =
+          new Invoke(buffer, "getHeapMemory", "heapBuffer", PRIMITIVE_BYTE_ARRAY_TYPE);
+      expressions.add(heapBuffer);
       ListExpression groupExpressions = new ListExpression();
       Expression readerAddr =
           new Invoke(buffer, "getUnsafeReaderAddress", "readerAddr", PRIMITIVE_LONG_TYPE);
-      // After this check, following read can be totally unsafe without checks.
-      ReplaceStub checkReadableBytesStub = new ReplaceStub();
-      expressions.add(checkReadableBytesStub);
       int acc = 0;
       boolean compressStarted = false;
       for (Descriptor descriptor : group) {

--- a/java/fury-core/src/main/java/org/apache/fury/io/FuryInputStream.java
+++ b/java/fury-core/src/main/java/org/apache/fury/io/FuryInputStream.java
@@ -76,15 +76,16 @@ public class FuryInputStream extends InputStream implements FuryStreamReader {
 
   private static byte[] growBuffer(int minFillSize, MemoryBuffer buffer) {
     int newSize;
-    int targetSize = buffer.size() + minFillSize;
+    int size = buffer.size();
+    int targetSize = size + minFillSize;
     newSize =
         targetSize < MemoryBuffer.BUFFER_GROW_STEP_THRESHOLD
             ? targetSize << 2
             : (int) Math.min(targetSize * 1.5d, Integer.MAX_VALUE - 8);
     byte[] newBuffer = new byte[newSize];
     byte[] heapMemory = buffer.getHeapMemory();
-    System.arraycopy(heapMemory, 0, newBuffer, 0, buffer.size());
-    buffer.initHeapBuffer(newBuffer, 0, buffer.size());
+    System.arraycopy(heapMemory, 0, newBuffer, 0, size);
+    buffer.initHeapBuffer(newBuffer, 0, size);
     heapMemory = newBuffer;
     return heapMemory;
   }

--- a/java/fury-core/src/test/java/org/apache/fury/StreamTest.java
+++ b/java/fury-core/src/test/java/org/apache/fury/StreamTest.java
@@ -35,6 +35,7 @@ import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.List;
 import org.apache.fury.config.CompatibleMode;
+import org.apache.fury.config.Language;
 import org.apache.fury.io.FuryInputStream;
 import org.apache.fury.io.FuryReadableChannel;
 import org.apache.fury.io.FuryStreamReader;
@@ -377,5 +378,27 @@ public class StreamTest extends FuryTestBase {
     assertEquals(fury.deserialize(stream), list);
     assertEquals(fury.deserialize(stream), new long[5000]);
     assertEquals(fury.deserialize(stream), new int[5000]);
+  }
+
+  public static class SimpleType {
+    public double dVal;
+
+    public SimpleType() {
+      dVal = 0.5;
+    }
+  }
+
+  // For issue https://github.com/apache/fury/issues/2060
+  @Test
+  public void testReadPrimitivesOnBufferFillBound() {
+    Fury fury = Fury.builder().withLanguage(Language.JAVA).build();
+    fury.register(SimpleType.class);
+    SimpleType v = new SimpleType();
+    ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+    fury.serialize(outputStream, v);
+    InputStream inputStream = new ByteArrayInputStream(outputStream.toByteArray());
+    FuryInputStream input = new FuryInputStream(inputStream, 11);
+    SimpleType newValue = (SimpleType) fury.deserialize(input);
+    Assert.assertEquals(v.dVal, newValue.dVal, 0.001);
   }
 }

